### PR TITLE
fix(oci-distribution): return content of all layers

### LINKED ISSUE #431: fix: oci-distribution overwrites content in previous layers with the last layer
If `oci-distribution` is to support pulling manifests containing more than one layer, we will need to ensure that each layer's content is returned.

https://github.com/deislabs/krustlet/blob/b5674ecc5d22ba95c6604fd96d8cb86c9a263f04/crates/oci-distribution/src/client.rs#L107-L115

### PR DESCRIPTION
**Pull request by bacongobbler**
closes #431

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

### CODE REVIEW

**Review by thomastaylor312 — APPROVED** [review]
One comment around how we are storing the layers I wanted to discuss, but nothing blocking

File: crates/oci-distribution/src/client.rs [hunk 1]
```diff
@@ -22,8 +22,8 @@ use www_authenticate::{Challenge, ChallengeFields, RawChallenge, WwwAuthenticate
 /// The data for an image or module.
 #[derive(Clone)]
 pub struct ImageData {
-    /// The content of the image or module.
-    pub content: Vec<u8>,
+    /// The layers of the image or module.
+    pub layers: Vec<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (bacongobbler) [reply crates/oci-distribution/src/client.rs:26]: Looking at this a little closer... It's unclear to me why we need an `ImageData` object that stores the layer content even though this information is already present in the `OciManifest` object.

I think we need to evaluate whether we can rip out `ImageData` and replace it with `OciManifest`, as the layer information is available in that object (`ImageData` appears to be a subset of `OciManifest`.

https://github.com/deislabs/krustlet/blob/b9dcbf2907778a68b34111a2d2935c4c2c946b60/crates/oci-distribution/src/client.rs#L22-L29
https://github.com/deislabs/krustlet/blob/b9dcbf2907778a68b34111a2d2935c4c2c946b60/crates/oci-distribution/src/manifest.rs#L25-L62

File: crates/kubelet/src/store/oci/file.rs [hunk 2]
```diff
@@ -84,8 +84,12 @@ impl Storer for FileStorer {
         if digest_path.exists() {
             tokio::fs::remove_file(&digest_path).await?;
         }
+        // FIXME: we need to determine the proper file path for each layer rather than assuming it's a single-layer image.
```
[the remark this answers is not in the document]
Comment (bacongobbler) [reply crates/kubelet/src/store/oci/file.rs:87]: https://github.com/deislabs/krustlet/issues/430 covers this in a sense

File: crates/kubelet/src/store/oci/file.rs [hunk 3]
```diff
@@ -84,8 +84,12 @@ impl Storer for FileStorer {
...
         let module_path = self.pull_file_path(image_ref);
-        tokio::fs::write(&module_path, image_data.content).await?;
+        if image_data.layers.is_empty() {
+            return Err(anyhow::anyhow!("No module layer present in image data"));
+        }
+        tokio::fs::write(&module_path, &image_data.layers[0]).await?;
```
[the remark this answers is not in the document]
Comment (bacongobbler) [reply crates/kubelet/src/store/oci/file.rs:92]: In a WebAssembly module, there should only be one layer. The old code in oci-distribution made the assumption that the last layer in the vector would be used as the webassembly module. This of course is not optimal and can lead to many frail cases. @itowlson and I discussed options last week and decided that it'd be best to tackle this in #430 so we can remove these assumptions.

This is also exasperated by the fact that we cannot fetch the layer's media type to determine whether the layer's content is actually a webassembly module or not, which I touched on a bit in https://github.com/deislabs/krustlet/pull/432#discussion_r512063196.

Really this whole section needs a rework... This was all written back to get things in a demo state, but there are some assumptions in here that need addressing.

**Pull request merged**